WordPress/gutenberg
# Layout: Reduce specificity of flex and grid margin overrides

⚠️ 🚧 🚧 🚧 🚧 This currently does not play nicely with Classic themes 🚧 🚧 🚧 🚧

In Classic themes, the `classic.css` file includes `margin-left: auto` and `margin-right: auto` rules. Reducing the specificity of the margin overrides for flex rules causes an issue in the block editor for the social icons block — the individual social icons blocks with this PR applied have margin rules applied that override the now reduced rule. Here's a screenshot of how it currently looks:

![image](https://github.com/WordPress/gutenberg/assets/14988353/78840155-d162-49cd-94a6-3004a6bb3ca5)

---

## What?

Reduce the specificity of the `flex` and `grid` layout type margin overrides, for consistency with the `flow` and `constrained` layout types.

This resolves an issue where flex and grid layout types have margin override specificity that interferes with setting margins at the block level in global styles.

## Why?


As reported in https://github.com/WordPress/gutenberg/pull/45196#pullrequestreview-1435758566 the flow and constrained layouts currently allow block-level-in-global-styles rules to override general layout styles. However, `flex` and `grid` don't have that lower level of specificity for their `margin: 0` rules. From a little digging, it appears that the solution back in https://github.com/WordPress/gutenberg/pull/47858 was to reduce specificity for the `spacingStyles` rules for each layout type. However, the `flex` and `grid` layouts had their overrides in `baseStyles`, which doesn't include the extra `:where()` rule to lower the specificity. The fix is to move the margin overrides for these two layout types over to `spacingStyles`.

## How?


* Move the `margin: 0` override rules for `flex` and `grid` layout types over to `spacingStyles` so that they receive the lower specificity rules introduced in #47858.

The difference in output of rules is:

```
body .is-layout-flex > * {
```

Becomes:

```
:where(body .is-layout-flex) > * {
```

## Testing Instructions





1. In the site editor, set the Cover block to have margin at the block level in global styles, e.g. set it to `50px`
2. Add a Stack block somewhere in your template.
3. Add a Cover block within the Stack block.
4. Note that the margins should appear to take effect in the site editor.
5. Save the template and global styles and view the site front end.
6. The margins should be consistent on the site front end.
7. Back in the site editor, add blocks before and after the Cover block, still within the Stack block. The margins should still take effect.
8. Save and double check on the site frontend.
9. Try the above, but activate the Grid experiment in Gutenberg and add the Cover block as a child of the Grid block. As with the Stack block, the margin rules should apply as expected in the editor and on the site frontend.
10. Double-check that default / fallback gap styles are still working as expected on Classic themes (e.g. test Columns and Social Icons blocks)

A question / note to the reviewer: the `flex` and `grid` rules, in contrast to flow/constrained, do not do anything about resetting margins for first/last items. I think that's likely preferable due to how these layout types work, but I'm curious what folks think. Is the change in this PR the desired approach for these layout types?

### Testing Instructions for Keyboard


## Screenshots or screencast 

Prior to this PR, the margins would only look correct in the site editor, but not on the site frontend. So for these before/after screenshots, we're looking at the site front end:

| Before | After |
| --- | --- |
| ![image](https://github.com/WordPress/gutenberg/assets/14988353/4d5ac779-2549-4689-8d6f-476c1263e3c7) | ![image](https://github.com/WordPress/gutenberg/assets/14988353/5f48bd1e-4e43-42fb-8fa0-ec9ac9b746d8) |
| ![image](https://github.com/WordPress/gutenberg/assets/14988353/f71b689b-2f62-4ca5-885c-0e6a29566013) | ![image](https://github.com/WordPress/gutenberg/assets/14988353/95ccf321-63af-4448-b98f-366f02f2ab03) |
